evm/p2p/peer.py

```diff
@@ -186,6 +190,21 @@ def send_hello(self):
         header, body = self.base_protocol.get_hello_message()
         self.send(header, body)
 
+    def match_protocols(self, remote_capabilities):
+        matching_capabilities = set(self.capabilities).intersection(remote_capabilities)
+        # Doing this on the sorted list of matching capabilities will give us a dict containing
+        # just the higher version of each matching protocol.
+        higher_matching = dict(
+            (name, version) for name, version in sorted(matching_capabilities))
```

Comment:
Experimenting with what this looks like to do using groupings instead of what feels a little *magical* relying on the way that dictionaries initialize when provided with multiple of the same key.

This example uses [`reduceby`](http://toolz.readthedocs.io/en/latest/api.html#toolz.itertoolz.reduceby) which is pretty much exactly what we want, but it could be expanded into multiple steps for readability.

```python
from cytoolz import reduceby, identity
higher_matching = reduceby(
    key=identity, 
    binop=lambda a, b: a if a.version > b.version else b,
    seq=matching_capabilities,
)
```

Reply:
That doesn't work when there are more than 2 versions of a given sub protocol. I could try and figure out how to make it work, but then again everyone reading this code would have to do the same if they wanted to make sure it does what it's expected. The dict() initialization trick, on the other hand, is probably going to be immediately understood by anybody familiar with Python (and there's the comment as well), so I think it'd be better to stick with the stdlib in this case. What do you say?

Reply:
> That doesn't work when there are more than 2 versions of a given sub protocol.

That is handled by `binop=lambda a, b: a if a.version > b.version else b`.

I can see where you are coming from as well.  I'm always a bit wary of code that takes advantage of a known behavior that isn't a formal part of the actual spec, but I think this is a mild enough version of that I can live with it.  Your call.

Reply:
Well, when I tried the reduceby variation as you pasted here, my unit test failed because there were multiple versions of a given protocol in the matching list when there should be only one. Now I can see it didn't work as expected because your code uses a.version and b.version in the binop comparison, but the items in matching_capabilities are two-tuples, which makes me wonder why that failed silently instead of raising an exception. Any ideas?

Reply:
Even when I change binop to compare the second element of the two-tuple, it still doesn't seem to do what we want:

```
In [8]: reduceby(
   ...:     key=identity,
   ...:     binop=lambda a, b: a if a[1] > b[1] else b,
   ...:     seq=sorted([('les', 1), ('les', 2)]))
   ...:
   ...:
Out[8]: {('les', 1): ('les', 1), ('les', 2): ('les', 2)}
```

Reply:
your key needs to be `operator.itemgetter(0)` to group them by the protocol name.

Reply:
Oh, right!
With this, I think I've addressed all the points you've made. Can you have another look?